### ISSUE TITLE
audit: "hook matcher never fires" is a false positive on regex matchers — it flags a working one and accepts a dead one

### ISSUE THREAD

**Issue by zernie** [zernie/agent-lint#131]
> **Rewritten 2026-08-09 after an experiment overturned the original report.** The first version of this issue said the rule had *caught* a real dead hook and merely gave bad remediation advice. That was wrong, and I had already changed a working matcher on its say-so. The rule is a **false positive**, which makes this more serious, not less. The original framing is preserved at the bottom.

## The experiment

Claude Code hook matchers are **regexes**. `mcp__.*` therefore matches `mcp__Google_Calendar__list_events` and the hook fires. Measured directly rather than argued:

```console
$ rm -f .claude/.cal-last-sync                 # remove the artifact the hook writes
# settings.json → PostToolUse matcher = "mcp__.*"
# … then one real Google Calendar MCP call in a live session …
$ ls .claude/.cal-last-sync && date -d @$(cat .claude/.cal-last-sync)
✅ written — 22:02:55 UTC
```

The hook fired. `vigiles audit` reports that same matcher as:

```
Hook matchers that never fire (1):
  ✗ Hook matcher "mcp__.*" is not a valid MCP tool reference
    (requires double underscores: `mcp__server__tool`). Did you mean "mcp__.*__.*"?
```

It does fire. The rule validates the matcher as if it were a literal tool name; matchers are patterns.

## Why this is worse than a nuisance

Following the finding makes the setup **worse**, in two ways.

**a. The suggestion never converges.** Apply `mcp__.*__.*` → the rule rejects its own suggestion and proposes `mcp__.*__.*__.*`. Each round appends `__.*`.

**b. The form it accepts is the one that is actually dead.** Four matchers, same repo, same run:

| matcher | rule verdict | fires on `mcp__Google_Calendar__list_events`? |
|---|---|---|
| `mcp__Google_Calendar__list_events` | accepted | yes (literal) |
| `mcp__[^_]+__[^_]+` | **accepted** | **no** |
| `mcp__.*__.*` | rejected | yes |
| `mcp__.*` | **rejected** | **yes — verified above** |
| `mcp__\w+__\w+` | rejected | no |

`[^_]+` cannot cross the underscore inside the server segment, and real server names contain underscores — `Google_Calendar` is Anthropic's own connector naming. So the rule **accepts a matcher that never fires and rejects one that does**, which is its verdict inverted on both rows that matter.

A user obeying the linter lands on a hook that lints clean and silently never runs. That is precisely the failure the rule exists to prevent, reached by complying with it.

## Why a pattern and not a literal

The server segment is not stable: the same Google Calendar server appears as `mcp__Google_Calendar__list_events` in one session and `mcp__4f54037d-0499-426a-8573-6130f3da1ef8__list_events` in another. A hook keyed to one literal id dies silently when the id changes. Requiring literals would make every MCP hook fragile in exactly that way.

## Suggested direction

Validate the matcher as a **pattern that can produce** an MCP tool name, not as a string that must look like one:

- test it against a synthetic `mcp__srv__tool` **and** against `mcp__some_server__do_thing` (underscore inside the server segment), and require both;
- that accepts `mcp__.*`, `mcp__.*__.*` and the literal form, and rejects `mcp__[^_]+__[^_]+` and `mcp__\w+__\w+` — the correct ranking, and the opposite of today's on the two contested rows;
- add a property test asserting that **the `Did you mean` suggestion satisfies the rule that produced it**. That alone would have caught the regress.

If the intent is narrower — flagging a matcher that cannot match *anything* rather than one that is imprecise — the finding text should say so, because "never fire" is what made me replace a working matcher.

## Cost here

One working matcher replaced with a narrower one on the strength of a false finding; one commit and one ledger entry written on the wrong premise and since corrected; a test written to enforce the rule, which would have rejected a valid matcher, caught only by mutation.

<details>
<summary>Original report (kept — its second and third points still stand)</summary>

The original issue reported the suggestion regress and the accept/reject inversion, but framed the rule as having correctly caught a dead hook. It had not; the hook was working. Points 1 and 2 above are unchanged from that report.
</details>
